test/replication-luatest/bootstrap_strategy_test.lua

```diff
@@ -179,3 +179,210 @@ g_auto.test_sync_waits_for_all_connected = function(cg)
+end)
+
+g_config.test_no_leader = function(cg)
+    cg.replica_set:start{wait_until_ready = false}
+    local logfile = fio.pathjoin(cg.server1.workdir, 'server1.log')
+    local empty_leader_msg = "the option can't be empty when bootstrap " ..
+                             "strategy is 'config'"
+    t.helpers.retrying({}, function()
+        t.assert(server:grep_log(empty_leader_msg, nil, {filename = logfile}))
+    end)
+end
+
+g_config.before_test('test_single_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            replication_timeout = 0.1,
+            bootstrap_strategy = 'config',
+            bootstrap_leader = server.build_listen_uri('server1'),
+            replication = server.build_listen_uri('server1'),
+        },
+    }
+end)
+
+g_config.after_test('test_single_leader', function(cg)
+    cg.server1:stop()
+end)
+
+g_config.test_single_leader = function(cg)
+    cg.replica_set:start()
+    cg.server1:exec(function()
+        local t = require('luatest')
+        t.assert_equals(box.info.status, 'running', 'server is working')
+    end)
+end
+
+local g_config_success = t.group('gh-7999-bootstrap-strategy-config-success', {
+                       {leader = 'server3'},
+                       {leader = uuid3},
+                   })
+
+g_config_success.before_test('test_correct_bootstrap_leader', function(cg)
+    local leader = cg.params.leader
+    -- cg.params can't have "/" for some reason, so recreate the path here.
+    if string.match(leader, 'server3') then
+        leader = server.build_listen_uri(leader)
+    end
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            bootstrap_strategy = 'config',
+            bootstrap_leader = leader,
+            instance_uuid = uuid1,
+            replication = {
+                server.build_listen_uri('server1'),
+                server.build_listen_uri('server2'),
+                server.build_listen_uri('server3'),
+            },
+            replication_timeout = 0.1,
+        },
+    }
+    cg.server2 = cg.replica_set:build_and_add_server{
+        alias = 'server2',
+        box_cfg = {
+            replicaset_uuid = 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
+            instance_uuid = uuid2,
+        }
+    }
+    cg.server3 = cg.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = {
+            replicaset_uuid = 'bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb',
+            instance_uuid = uuid3,
+            listen = {
+                server.build_listen_uri('server3'),
+                server.build_listen_uri('server3-alias')
+            },
+        },
+    }
+end)
+
+g_config_success.after_each(function(cg)
+    cg.replica_set:drop()
+end)
+
+g_config_success.test_correct_bootstrap_leader = function(cg)
+    cg.replica_set:start{}
```

Comment:
Maybe we should add a test that server can bootstrap in `config` mode even if it can only connect to the bootstrap leader, while having many (eg 3 or 5) servers listed in `replication`?

Reply:
Good idea, done:
```diff
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index b1b95a7ca..a542b23f4 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -8,6 +8,9 @@ local g_auto = t.group('gh-5272-bootstrap-strategy-auto')
 local uuid1 = '11111111-1111-1111-1111-111111111111'
 local uuid2 = '22222222-2222-2222-2222-222222222222'
 local uuid3 = '33333333-3333-3333-3333-333333333333'
+local uuida = 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa'
+local uuidb = 'bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb'
+local uuidc = 'cccccccc-cccc-cccc-cccc-cccccccccccc'
 
 local function server_is_ready(server)
     local ok, _ = pcall(function()
@@ -222,10 +225,6 @@ g_config.before_test('test_uuid', function(cg)
```
<skipped some moving of `after_test`>
```diff
@@ -327,23 +326,34 @@ g_config.test_single_leader = function(cg)
     end)
 end
 
+g_config.after_test('test_single_leader', function(cg)
+    cg.server1:stop()
+end)
+
 local g_config_success = t.group('gh-7999-bootstrap-strategy-config-success', {
                        {leader = 'server3'},
                        {leader = uuid3},
                    })
 
-g_config_success.before_test('test_correct_bootstrap_leader', function(cg)
-    local leader = cg.params.leader
+g_config_success.before_each(function(cg)
+    cg.leader = cg.params.leader
     -- cg.params can't have "/" for some reason, so recreate the path here.
-    if string.match(leader, 'server3') then
-        leader = server.build_listen_uri(leader)
+    if string.match(cg.leader, 'server3') then
+        cg.leader = server.build_listen_uri(cg.leader)
     end
+end)
+
+g_config_success.after_each(function(cg)
+    cg.replica_set:drop()
+end)
+
+g_config_success.before_test('test_correct_bootstrap_leader', function(cg)
     cg.replica_set = replica_set:new{}
     cg.server1 = cg.replica_set:build_and_add_server{
         alias = 'server1',
         box_cfg = {
             bootstrap_strategy = 'config',
-            bootstrap_leader = leader,
+            bootstrap_leader = cg.leader,
             instance_uuid = uuid1,
             replication = {
                 server.build_listen_uri('server1'),
@@ -356,33 +366,65 @@ g_config_success.before_test('test_correct_bootstrap_leader', function(cg)
     cg.server2 = cg.replica_set:build_and_add_server{
         alias = 'server2',
         box_cfg = {
-            replicaset_uuid = 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
+            replicaset_uuid = uuida,
             instance_uuid = uuid2,
         }
     }
     cg.server3 = cg.replica_set:build_and_add_server{
         alias = 'server3',
         box_cfg = {
-            replicaset_uuid = 'bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb',
+            replicaset_uuid = uuidb,
             instance_uuid = uuid3,
             listen = {
                 server.build_listen_uri('server3'),
-                server.build_listen_uri('server3-alias')
             },
         },
     }
 end)
 
-g_config_success.after_each(function(cg)
-    cg.replica_set:drop()
-end)
-
 g_config_success.test_correct_bootstrap_leader = function(cg)
     cg.replica_set:start{}
-    t.helpers.retrying({}, cg.server1.exec, cg.server1, function()
+    t.helpers.retrying({}, cg.server1.exec, cg.server1, function(uuid)
         local t = require('luatest')
-        t.assert_equals(box.info.cluster.uuid,
-                        'bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb',
+        t.assert_equals(box.info.cluster.uuid, uuid,
                         'Server bootstrapped from correct leader')
-    end)
+    end, {uuidb})
+end
+
+g_config_success.before_test('test_wait_only_for_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            bootstrap_strategy = 'config',
+            bootstrap_leader = cg.leader,
+            replication = {
+                server.build_listen_uri('server1'),
+                server.build_listen_uri('unreachable_2'),
+                server.build_listen_uri('server3'),
+                server.build_listen_uri('unreachable_4'),
+            },
+            replication_connect_timeout = 1000,
+            replication_timeout = 0.1,
+        },
+    }
+    cg.server3 = cg.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = {
+            replicaset_uuid = uuidb,
+            instance_uuid = uuid3,
+            listen = {
+                server.build_listen_uri('server3'),
+            },
+        },
+    }
+end)
+
+g_config_success.test_wait_only_for_leader = function(cg)
+    cg.replica_set:start{}
+    t.helpers.retrying({}, cg.server1.exec, cg.server1, function(uuid)
+        local t = require('luatest')
+        t.assert_equals(box.info.cluster.uuid, uuid,
+                        'Server boots as soon as sees the leader')
+    end, {uuidb})
 end

```